### ISSUE TITLE
Test failure "testEval" with Redis 5.0RC6

### ISSUE THREAD

**Issue by remicollet** [phpredis/phpredis#1424]
```
testEval                                                       [FAILED]

Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4526 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4563 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)
Assertion failed: /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:4582 (testEval)

```

**Comment by remicollet** [comment]
Everything OK with redis 4.0.11

**Comment by michael-grunder** [comment]
Hey,

I wasn't able to replicate this failure.  The test is passing for me on my MBP and Ubuntu box (and it's passing in Travis).

What OS/version of Redis?

**Comment by remicollet** [comment]
> What OS/version of Redis?

As in the tile "5.0RC5" (on Fedora 27)

**Comment by remicollet** [comment]
Notice 4.1.1 also fails

**Comment by yatsukhnenko** [comment]
`return redis.call('smembers', KEYS[1])` returns set with wrong order

**Comment by yatsukhnenko** [comment]
And even `$this->redis->smembers('{eval-key}-set');` returns wrong order.

**Comment by yatsukhnenko** [comment]
Redis Sets are **unordered** collections of strings.
So probably there is something wrong in `array_diff_recursive`

**Comment by yatsukhnenko** [comment]
I mean `array_diff_recursive` shouldn't be used to compare results in testEval

**Comment by michael-grunder** [comment]
I changed our EVAL test to use a list and a zset since those have a deterministic order in 2e412373c47a574a1215756e79e6512a46e30cdc.

Also sorry for asking what version of Redis when you mentioned that _in the title_ :smile:

**Comment by michael-grunder** [comment]
@remicollet Is this fixed for you?

**Comment by remicollet** [comment]
Indeed, ok with  RC1 + 2e41237.

```
testEval                                                       [PASSED]
testEvalSHA                                                    [PASSED]

```

Will dig later (next week) on new failure

```
testXRange                                                     
Warning: Invalid argument supplied for foreach() in /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php on line 5318
[FAILED]
testXGroup                                                     [FAILED]
testXAck                                                       [FAILED]
testXRead                                                      [FAILED]
testXReadGroup                                                 [FAILED]
testXPending                                                   [FAILED]
testXDel                                                       [FAILED]
testXTrim                                                      [FAILED]
testXClaim                                                     [FAILED]

```
And
```
Uncaught exception 'protocol error, got '1' as reply type byte
' (testXAdd)
Assertion failed (1 !== 3): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5315 (doXRangeTest)
Assertion failed (1 !== 2): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5327 (doXRangeTest)
Assertion failed (1 !== 3): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5327 (doXRangeTest)
Uncaught exception 'protocol error, got '1' as reply type byte
' (testXRange)
Uncaught exception 'protocol error, got '5' as reply-type byte
' (testXGroup)
Uncaught exception 'protocol error, got '4' as reply-type byte
' (testXAck)
Assertion failed ( !== Array
(
    [{stream}-1] => Array
        (
            [0] => Array
                (
                    [f1] => v1
                    [f2] => v2
                )

        )

    [{stream}-2] => Array
        (
            [0] => Array
                (
                    [f1] => v1
                    [f2] => v2
                )

        )

)
): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5425 (doXReadTest)
Assertion failed (0 !== 1): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5431 (doXReadTest)
Assertion failed (0 !== 1): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5431 (doXReadTest)
Assertion failed (0 !== 2): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5431 (doXReadTest)
Assertion failed (0 !== 2): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5431 (doXReadTest)
Assertion failed (1 !== 0): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5436 (doXReadTest)
Assertion failed ( !== Array
(
    [{stream}-1] => Array
        (
            [0] => Array
                (
                    [final] => row
                )

        )

)
): /dev/shm/BUILD/php-pecl-redis4-4.2.0~RC1/NTS/tests/RedisTest.php:5444 (doXReadTest)
Uncaught exception 'protocol error, got '1' as reply type byte
' (testXRead)
Uncaught exception 'protocol error, got '4' as reply-type byte
' (testXReadGroup)
Uncaught exception 'protocol error, got '0' as reply-type byte
' (testXPending)
Uncaught exception 'protocol error, got '0' as reply type byte
' (testXDel)
Uncaught exception 'protocol error, got '2' as reply type byte
' (testXTrim)
Uncaught exception 'protocol error, got '2' as reply-type byte
' (testXClaim)

```

**Comment by michael-grunder** [comment]
The STREAMS failure may be related to a last minute change in Redis.  The day before 5.0 GA was published he switched from using STATUS to BULK replies for the stream IDs.

The phpredis commit for that 0b97ec3739d99f0778ff827cb58c011b92d27a74 is in the develop branch but not  2e41237 so you may be hitting an incompatibility there.

**Comment by remicollet** [comment]
Seems OK using RC1 +  0b97ec3 +  2e41237 
`
All tests passed. \o/
`

**Comment by yatsukhnenko** [comment]
Yes, at the end of this week

**Comment by yatsukhnenko** [comment]
Fixed in 4.2.0
